### PR TITLE
Implement RAII wrapper for ctx_* functions

### PR DESCRIPTION
**Pull request by kinkie**
Current uses of ctx_enter and ctx_exit are scope-based.
We can enrich them by implementing a RAII wrapper class
and fully rely on scope for entering and leaving contexts safely

### CODE REVIEW

File: src/DebugContext.h
Comment (yadij) [inline src/DebugContext.h:23]: Please follow squid style guidelines for class layout:

```suggestion
class DebugContext
{
public:
    DebugContext(const char *descr) {
        context_ = ctx_enter(descr);
    };
    ~DebugContext() {
        ctx_exit(context_);
    }

private:
    Ctx context_;
};
```

File: src/Makefile.am
Comment (yadij) [inline src/Makefile.am:840]: Unrelated formatting change. Please revert.
```suggestion
# stock tools for unit tests - library independent versions of dlink_list 
```

File: src/Makefile.am
Comment (yadij) [inline src/Makefile.am:762]: Unrelated formatting change. Please revert.
```suggestion
# The cf_gen_defines.cci is auto-generated and does not exist when the 
```

File: src/http.cc
Comment (yadij) [inline src/http.cc:665]: Why the mix between "DebugContext ctx(..." and "auto ctx = DebugContext(..." ?

Please try to make them all use one style if possible and if you have to pick the second, document why in the PR description.

**Review by yadij — CHANGES_REQUESTED** [review]

**Comment by yadij** [comment]
It is too much to ask to move the ctx_enter/exit to be private members of this class and use it as the only way to do this type of debugging?

File: src/DebugContext.h
Comment (kinkie) [reply src/DebugContext.h:23]: ok, thanks. I'm rusty :(

File: src/Makefile.am
Comment (kinkie) [reply src/Makefile.am]: ok. I'll run a sourceformat before merging just to be on the safe side

File: src/Makefile.am
Comment (kinkie) [reply src/Makefile.am]: ok

File: src/http.cc
Comment (kinkie) [reply src/http.cc:665]: Switching all to the first form

**Comment by kinkie** [comment]
> It is too much to ask to move the ctx_enter/exit to be private members of this class and use it as the only way to do this type of debugging?

Ok

**Comment by kinkie** [comment]
> > It is too much to ask to move the ctx_enter/exit to be private members of this class and use it as the only way to do this type of debugging?
> 
> Ok

Checking it, it is quite complicated and would bring relatively little added benefits. They bring in a bunch of static variables and other functions, and there is some gnarly interaction between ctx_*, ctx_print, db_print and its private implementations.
If you want I'll go ahead, but it's serious-er code surgery.

**Review by rousskov — CHANGES_REQUESTED** [review]
Please keep this soon-to-be-removed code as-is or remove the entire legacy feature instead.

Please see https://github.com/measurement-factory/squid/commit/944cb4c4152db4a97c8e0ec995ea89efb757f785 for rationale (and implementation). We had not posted that commit for the official review because I wanted to improve modern CodeContext coverage first. However, modern coverage is already much better than this legacy code, so the legacy code can be removed now, despite a few known holes in CodeContext coverage.

**Comment by kinkie** [comment]
Given @rousskov's comment, are we okay with a lightweight wrapper? I've implemented the other changes

**Comment by rousskov** [comment]
> Given @rousskov's comment, are we okay with a lightweight wrapper?

No, we are not. We should not waste more time on the code that should be removed. I cannot give you back the time you have wasted already, unfortunately. I can only prevent further waste (yours, mine, and others).

**Review by rousskov — CHANGES_REQUESTED** [review]
I am not sure why my review was requested after I have just submitted one. There have been no changes in my position in the past few minutes.

**Comment by kinkie** [comment]
On Sat, 22 May 2021 at 15:41, Alex Rousskov ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> I am not sure why my review was requested after I have just submitted one.
> There have been no changes in my position in the past few minutes
>

I had interpreted your comment as only applying to the scope expansion
suggested by Amos.
If you are blocking the whole PR and request it to be abandoned, please be
more explicit about that.
I’ll cancel the PR.

> --
@mobile

**Comment by kinkie** [comment]
abandoning due to more comprehensive measurement-factory project

**Comment by rousskov** [comment]
> If you are blocking the whole PR and request it to be abandoned, please be more explicit about that.

* If the above was meant as a critique of my past comments (as in "they were not explicit enough"), then I do not think it is a reasonable assessment because I was not responding to Amos or commenting on his review in any way. I was reviewing your PR. One should not add "I am reviewing this PR independently from whatever other reviewers are doing" disclaimer to every review. It is the _default_. This is especially true for reviews submitted a few hours apart from each other and from PR opening.
 
* If the above was meant as a clarification request for going forward, then, yes, I thought that this PR should be either refactored to completely remove the legacy feature or closed without changes (in anticipation of a PR that will remove the legacy feature). I am not sure how I could be more explicit about those two options though.

> abandoning due to more comprehensive measurement-factory project

Just to avoid a misunderstanding: There is no "more comprehensive measurement-factory project" that blocked this PR! The CodeContext feature that I have mentioned in my [review](https://github.com/squid-cache/squid/pull/828#pullrequestreview-666182263) is an _official_ feature. The core of that work has been committed in 2019 (ccfbe8f). The unofficial Factory commit (https://github.com/measurement-factory/squid/commit/944cb4c4152db4a97c8e0ec995ea89efb757f785) that I mentioned in my review simply removes the legacy feature that this PR is improving (and also provides rationale for that removal -- which is the reason I mentioned that unofficial commit).

If you would like to remove this legacy feature, please feel free to do so! You can reopen this PR or open another one for that.

**Comment by kinkie** [comment]
On Sat, May 22, 2021 at 8:44 PM Alex Rousskov ***@***.***>
wrote:

> If you are blocking the whole PR and request it to be abandoned, please be
> more explicit about that.
>
>
>    -
>
>    If the above was meant as a critique of my past comments (as in "they
>    were not explicit enough"), then I do not think it is a reasonable
>    assessment because I was not responding to Amos or commenting on his review
>    in any way. I was reviewing your PR. One should not add "I am reviewing
>    this PR independently from whatever other reviewers are doing" disclaimer
>    to every review. It is the *default*. This is especially true for
>    reviews submitted a few hours apart from each other and from PR opening.
>    -
>
>    If the above was meant as a clarification request for going forward,
>    then, yes, I thought that this PR should be either refactored to completely
>    remove the legacy feature or closed without changes (in anticipation of a
>    PR that will remove the legacy feature). I am not sure how I could be more
>    explicit about those two options though.
>
> abandoning due to more comprehensive measurement-factory project
>
> Just to avoid a misunderstanding: There is no "more comprehensive
> measurement-factory project" that blocked this PR! The CodeContext feature
> that I have mentioned in my review
> <https://github.com/squid-cache/squid/pull/828#pullrequestreview-666182263>
> is an *official* feature. The core of that work has been committed in
> 2019 (ccfbe8f
> <https://github.com/squid-cache/squid/commit/ccfbe8f49d36e8cce928b552a2ad7b4343ccdabd>).
> The unofficial Factory commit ***@***.***
> <https://github.com/measurement-factory/squid/commit/944cb4c4152db4a97c8e0ec995ea89efb757f785>)
> that I mentioned in my review simply removes the legacy feature that this
> PR is improving (and also provides rationale for that removal -- which is
> the reason I mentioned that unofficial commit).
>

No critique expressed or implied, I apologise if what I wrote could be
interpreted like that.
Got you; I'm just trying to find something to work on that is within my
skill range, useful and not already being worked on. So far I think i have
a 50% hit rate.


> If you would like to remove this legacy feature, please feel free to do
> so! You can reopen this PR or open another one for that.
>

If work is underway to do that, I'm not keen on duplicating that

-- 
    Francesco

**Comment by rousskov** [comment]
> I'm just trying to find something to work on that is within my skill range, useful and not already being worked on.

Do you think a list of straightforward but useful TODOs would help you find good candidates? If there is demand for it, I can start maintaining one in [squid-notes](https://github.com/measurement-factory/squid-notes/tree/start#readme).

> If work is underway to do that, I'm not keen on duplicating that

Your call, of course. It would probably save _my_ time if you port that Factory [commit](https://github.com/measurement-factory/squid/commit/944cb4c4152db4a97c8e0ec995ea89efb757f785) to master and guide it through the official review process, but that is obviously rather boring work, and I am not asking you to do it.

**Comment by kinkie** [comment]
On Sat, May 22, 2021 at 9:14 PM Alex Rousskov ***@***.***>
wrote:

> I'm just trying to find something to work on that is within my skill
> range, useful and not already being worked on.
>
> Do you think a list of straightforward but useful TODOs would help you
> find good candidates? If there is demand for it, I can start maintaining
> one in squid-notes
> <https://github.com/measurement-factory/squid-notes/tree/start#readme>.
>
sounds good, thanks for offering
-- 
    Francesco

**Comment by rousskov** [comment]
> sounds good

The [initial](https://github.com/measurement-factory/squid-notes/commit/a9f531c8edc3c501d7f8e2496d618eeda88c8127) list is at https://github.com/measurement-factory/squid-notes/blob/start/todos.md. I plan to add more items as time allows. Please let me know if you find any of the existing items useful or just have improvement suggestions, but we should probably move any followup discussion to email. Thank you.

**Pull request closed without merging**
